src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,54 @@
+#include "firmware/shared/circular_buffer.h"
+
+void circular_buffer_init(struct circular_buffer *cbuff)
+{
+    cbuff->head   = 0;
+    cbuff->tail   = 0;
+    cbuff->isFull = false;
+    // Initialize buffer size to CIRCULAR_BUFFER_MAX_NUM_ELEMENTS
+    cbuff->bufferData = (float *)malloc(sizeof(float) * CIRCULAR_BUFFER_MAX_NUM_ELEMENTS);
+}
+
+void circular_buffer_free(struct circular_buffer *cbuff)
+{
+    free(cbuff->bufferData);
+}
+
+// Note: Implementation assumes head and tail are equal when full
+void circular_buffer_push(struct circular_buffer *cbuff, float data)
+{
+    cbuff->bufferData[cbuff->head % CIRCULAR_BUFFER_MAX_NUM_ELEMENTS] = data;
+    cbuff->head++;
+    if (cbuff->head > CIRCULAR_BUFFER_MAX_NUM_ELEMENTS)
+    {
+        cbuff->tail++;
+    }
+    if (cbuff->head == cbuff->tail)
+    {
+        cbuff->isFull = true;
+    }
+}
+
+// Note: Index is starting from 1 for most recent
+float circular_buffer_get_index(struct circular_buffer *cbuff, int index)
+{
+    // Check if index is larger than the maximum buffer size
+    if (index > CIRCULAR_BUFFER_MAX_NUM_ELEMENTS)
+    {
+        return NAN;
+    }
+
+    // Check if enough items are in the buffer for index
+    if (cbuff->isFull == false && index > cbuff->head)
```

Comment:
If there is only 1 item in the buffer so far, and you want that item, passing in 1 returns a NAN instead of the item

Reply:
If there's 1 item, `head` would be pointing to 1 (it points to whatever it will be writing to next). If the `index` passed in is 1, the equality becomes 1 > 1 which is false. So it will still return the right value.

Reply:
Big rip, completely misread that statement. You right

Reply:
All good, I thought that was a problem too until I ran through it again.